### PR TITLE
RDBQA-82 Add option to run encrypted database  when authentication is off

### PR DESCRIPTION
**Pull request by itayweb**
### Issue link
https://issues.hibernatingrhinos.com/issue/RDBQA-82/Add-option-to-run-encrypted-database-when-authentication-is-off
### Additional description
Allow creating encrypted database without ssl server
### Type of change
- Bug fix
### How risky is the change?
- Low 
### Testing by RavenDB QA team
- This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.

### CODE REVIEW

**Review by garayx — APPROVED** [review]

File: scripts/buildProjects.ps1 [hunk 1]
```diff
@@ -32,6 +32,10 @@ function BuildServer ( $srcDir, $outDir, $target) {
         $commandArgs += "/p:UseAppHost=false"
     }
 
+    if ($env:ALLOW_ENCRYPTED_SSL_ONLY) {
```
Comment (ppekrol) [inline scripts/buildProjects.ps1]: Would prefer that to be just a switch in the build script

File: src/Raven.Server/Raven.Server.csproj [hunk 2]
```diff
@@ -240,4 +240,8 @@
     <ConcurrentGarbageCollection>false</ConcurrentGarbageCollection>
     <RetainVMGarbageCollection>false</RetainVMGarbageCollection>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(AllowEncryptedWithoutSsl)' == 'True'">
+    <DefineConstants>$(DefineConstants);ALLOW_ENCRYPTED_SSL_ONLY</DefineConstants>
+    <Optimize>true</Optimize>
```
Comment (ppekrol) [inline src/Raven.Server/Raven.Server.csproj]: this is not needed?

File: src/Raven.Server/Raven.Server.csproj [hunk 3]
```diff
@@ -240,4 +240,8 @@
     <ConcurrentGarbageCollection>false</ConcurrentGarbageCollection>
     <RetainVMGarbageCollection>false</RetainVMGarbageCollection>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(AllowEncryptedWithoutSsl)' == 'True'">
+    <DefineConstants>$(DefineConstants);ALLOW_ENCRYPTED_SSL_ONLY</DefineConstants>
```
Comment (ppekrol) [inline src/Raven.Server/Raven.Server.csproj]: let's be consistent `ALLOW_ENCRYPTED_WITHOUT_SSL`

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 4]
```diff
@@ -887,7 +887,11 @@ public RavenConfiguration CreateDatabaseConfiguration(StringSegment databaseName
                 var record = databaseRecord.MaterializedRecord;
                 if (record.Encrypted)
                 {
-                    if (_serverStore.Server.WebUrl?.StartsWith("https:", StringComparison.OrdinalIgnoreCase) == false)
+                    var allowEncryptedSslOnly = true;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: `var allowEncryptedSslOnly = true;` => `var allowEncryptedOverHttp = false;`

File: scripts/buildProjects.ps1 (hunk 1 above)
Comment (itayweb) [reply scripts/buildProjects.ps1]: Done

File: src/Raven.Server/Raven.Server.csproj (hunk 2 above)
Comment (itayweb) [reply src/Raven.Server/Raven.Server.csproj]: Done

File: src/Raven.Server/Raven.Server.csproj (hunk 3 above)
Comment (itayweb) [reply src/Raven.Server/Raven.Server.csproj]: Done

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 4 above)
Comment (itayweb) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: Done

File: build.ps1 [hunk 5]
```diff
@@ -15,7 +15,8 @@ param(
     [switch]$NoBundling,
     [switch]$DryRunVersionBump = $false,
     [switch]$DryRunSign = $false,
-    [switch]$Help)
+    [switch]$Help,
```
Comment (ppekrol) [inline build.ps1]: usually help is last

Comment (itayweb) [reply build.ps1]: Done

File: build.ps1 [hunk 6]
```diff
@@ -15,6 +15,7 @@ param(
     [switch]$NoBundling,
     [switch]$DryRunVersionBump = $false,
     [switch]$DryRunSign = $false,
+    [switch]$AllowEncryptedOverHttp,
```
Comment (gregolsky) [inline build.ps1]: let's not add a swithc for that but use env var

File: scripts/buildProjects.ps1 [hunk 7]
```diff
@@ -1,5 +1,5 @@
 
-function BuildServer ( $srcDir, $outDir, $target) {
+function BuildServer ( $srcDir, $outDir, $target, $ALLOW_ENCRYPTED_OVER_HTTP) {
```
Comment (gregolsky) [inline scripts/buildProjects.ps1]:
```suggestion
function BuildServer ( $srcDir, $outDir, $target) {
```

File: scripts/buildProjects.ps1 [hunk 8]
```diff
@@ -32,6 +32,10 @@ function BuildServer ( $srcDir, $outDir, $target) {
         $commandArgs += "/p:UseAppHost=false"
     }
 
+    if ($ALLOW_ENCRYPTED_OVER_HTTP) {
```
Comment (gregolsky) [inline scripts/buildProjects.ps1]:
```suggestion
    if ($env:RAVEN_BuildOptions -like "*AllowEncryptedOverHttp*") {
```

File: scripts/archive.ps1 [hunk 9]
```diff
@@ -35,41 +35,56 @@ function UnpackTarGzToDir ( $archivePath, $outDir ) {
...
         $7za = [io.path]::combine("scripts", "assets", "bin", "7za.exe")
         & "$7za" a -ttar "$targetFilename.tar" $glob
         CheckLastExitCode
-        & "$7za" a -tbzip2 "$targetFilename.tar.bz2" "$targetFilename.tar"
+        if ($ALLOW_ENCRYPTED_OVER_HTTP -and !($string -contains "Tools") ){
+            & "$7za" a -tbzip2 "$targetFilename-encrypted-over-http.tar.bz2" "$targetFilename.tar"
+        }
+        else {
+            & "$7za" a -tbzip2 "$targetFilename.tar.bz2" "$targetFilename.tar"
+        }
```
Comment (gregolsky) [inline scripts/archive.ps1]: This is not the responsibility of this script to name the files according to your needs. You need to perform these operations in your wrapper script or manually.

Please revert changes in this file.

File: build.ps1 [hunk 10]
```diff
@@ -197,7 +202,7 @@ Foreach ($target in $targets) {
     $specOutDir = [io.path]::combine($OUT_DIR, $target.Name)
     CleanDir $specOutDir
 
-    BuildServer $SERVER_SRC_DIR $specOutDir $target
+    BuildServer $SERVER_SRC_DIR $specOutDir $target $ALLOW_ENCRYPTED_OVER_HTTP
```
Comment (gregolsky) [inline build.ps1]: Passing options that are not used very often down the whole pipeline creates a lot of clutter. Just use the env var check whenever you need to.

File: build.ps1 [hunk 11]
```diff
@@ -55,6 +56,10 @@ if ($Osx) {
 
 CheckPrerequisites
 
+if ($AllowEncryptedOverHttp) {
+    $ALLOW_ENCRYPTED_OVER_HTTP = $true;
+}
+
```
Comment (gregolsky) [inline build.ps1]: This won't be required if used env var

File: scripts/package.ps1 [hunk 12]
```diff
@@ -1,5 +1,5 @@
 
-function CreateServerPackage ( $projectDir, $releaseDir, $packOpts ) {
+function CreateServerPackage ( $projectDir, $releaseDir, $packOpts, $ALLOW_ENCRYPTED_OVER_HTTP ) {
```
Comment (gregolsky) [inline scripts/package.ps1]: Please revert these changes.

File: build.ps1 [hunk 13]
```diff
@@ -252,7 +257,7 @@ Foreach ($target in $targets) {
         }
     }
 
-    CreateServerPackage $PROJECT_DIR $RELEASE_DIR $packOpts
+    CreateServerPackage $PROJECT_DIR $RELEASE_DIR $packOpts $ALLOW_ENCRYPTED_OVER_HTTP
```
Comment (gregolsky) [inline build.ps1]: same thing

File: src/Raven.Server/Raven.Server.csproj [hunk 14]
```diff
@@ -240,4 +240,7 @@
     <ConcurrentGarbageCollection>false</ConcurrentGarbageCollection>
     <RetainVMGarbageCollection>false</RetainVMGarbageCollection>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(AllowEncryptedOverHttp)' == 'True'">
+    <DefineConstants>$(DefineConstants);ALLOW_ENCRYPTED_OVER_HTTP</DefineConstants>
+  </PropertyGroup>
```
Comment (gregolsky) [inline src/Raven.Server/Raven.Server.csproj]: You could actually generalize this whole PR into a few lines of csproj, no PS code modification whatsoever, by using the env var directly in here. 

See: https://learn.microsoft.com/en-us/visualstudio/msbuild/how-to-use-environment-variables-in-a-build?view=vs-2022

This is good because:
- you don't modify powershell pipeline and drop all the above changes
- you can easily add other constants, mix them etc.

However I think you'd rather after modifying *Directory.build.props* than just the server then.

Consider the following program:
```
#if (OPTION1)

Console.WriteLine("Option 1");

#endif

#if (OPTION2)

Console.WriteLine("Option 2");

#endif
```

and its project file
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

 <PropertyGroup Condition="'$(RAVEN_BuildOptions)' != ''">
    <DefineConstants>$(DefineConstants);$(RAVEN_BuildOptions)</DefineConstants>
  </PropertyGroup>

</Project>
```

And then this output:
```
C:\work\debug\buildopt> $env:RAVEN_BuildOptions='OPTION1'
C:\work\debug\buildopt> dotnet run
Option 1
C:\work\debug\buildopt> $env:RAVEN_BuildOptions='OPTION1,OPTION2'
C:\work\debug\buildopt> dotnet run
Option 1
Option 2
C:\work\debug\buildopt>
```

Ain't that cool?

File: scripts/buildProjects.ps1 [hunk 15]
```diff
@@ -32,6 +32,10 @@ function BuildServer ( $srcDir, $outDir, $target) {
...
+        $commandArgs += "/p:AllowEncryptedOverHttp=True"
```
Comment (gregolsky) [inline scripts/buildProjects.ps1]: Actually we don't need this at all. Please see my comment in csproj :)

**Review by gregolsky — CHANGES_REQUESTED** [review]

File: scripts/buildProjects.ps1 (hunk 15 above)
Comment (gregolsky) [reply scripts/buildProjects.ps1]: OK, we had a little talk with Itay and decided to handle it this way:
- handle the RAVEN_BuildOptions change in csproj
- add a string switch called "$BuildOptions" defaulting to empty string
- if that's not null or empty, set $env:RAVEN_BuildOptions $BuildOptions
- add a string switch called "$ArtifactNameSuffix" defaulting to empty string
-  if that's not null or empty, set $env:ArtifactNameSuffix = $ArtifactNameSuffix
- where the filenames are set, instead of a targeted `if` let's do e.g. `$zipFile = "$($targetFilename)$($env:ArtifactNameSuffix).zip"`

Comment (itayweb) [reply scripts/buildProjects.ps1]: Done

File: scripts/buildProjects.ps1 [hunk 16]
```diff
@@ -32,6 +32,10 @@ function BuildServer ( $srcDir, $outDir, $target) {
         $commandArgs += "/p:UseAppHost=false"
     }
 
+    if ($env:RAVEN_BuildOptions -like "*AllowEncryptedOverHttp*") {
```
Comment (gregolsky) [inline scripts/buildProjects.ps1]: You do not need this. It should be automaticall applied when you set the `$env:RAVEN_BuildOptions` using below csproj change

Comment (itayweb) [reply scripts/buildProjects.ps1]: Done

**Review by gregolsky — APPROVED** [review]

**Review by gregolsky — APPROVED** [review]

**Review by gregolsky — APPROVED** [review]

**Review by gregolsky — APPROVED** [review]

**Review by gregolsky — APPROVED** [review]
